src/SourceCode.Clay.Threading/AsyncLock.cs

```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SourceCode.Clay.Threading
+{
+    /// <summary>
+    /// Represents a lock that can operate under an asynchronous context.
+    /// </summary>
+    public sealed partial class AsyncLock : IDisposable
+    {
+        private readonly SemaphoreSlim _semaphore;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="AsyncLock"/> class.
+        /// </summary>
+        public AsyncLock()
+        {
+            _semaphore = new SemaphoreSlim(1);
+        }
+
+        internal void Release() => _semaphore.Release();
+
+        private AsyncLockCookie Wait(bool result)
+            => result
+                ? new AsyncLockCookie(this)
+                : new AsyncLockCookie();
+
+        private ValueTask<AsyncLockCookie> WaitAsync(Task<bool> wait)
+        {
+            async ValueTask<AsyncLockCookie> ImplAsync(Task<bool> wait)
+                => Wait(await wait);
+
+            if (wait.IsCompleted)
+                return new ValueTask<AsyncLockCookie>(Wait(wait.Result));
+            else
+                return ImplAsync(wait);
+        }
+
+        private bool TryAcquire(out ValueTask<AsyncLockCookie> result, bool ping, CancellationToken cancellationToken)
+        {
+            var acquired = _semaphore.Wait(0, cancellationToken);
```

Comment:
Prevents allocation of a Task or associated state machine if the semaphore can be acquired immediately.